pilot/cmd/pilot-agent/readiness.go

```diff
@@ -0,0 +1,153 @@
+		checkHasDynamicActiveListeners,
+	}
+)
+
+type checkFunc func(configDump *adminapi.ConfigDump) error
+
+// checkReady polls the Envoy admin port to determine if the Envoy instance has received its configuration from Pilot.
+func checkReady(adminPort int) (err error) {
+	// Add a prefix to any returned errors.
+	defer func() {
+		if err != nil {
+			err = multierror.Prefix(err, "Envoy proxy is NOT ready.")
+		}
+	}()
+
+	var configDump *adminapi.ConfigDump
+	configDump, err = getConfigDump(adminPort)
+	if err != nil {
+		return err
+	}
+
+	for _, f := range checkFuncs {
+		if e := f(configDump); e != nil {
+			err = multierror.Append(err, e)
+		}
+	}
+	return err
+}
+
+func checkHasDynamicActiveClusters(configDump *adminapi.ConfigDump) error {
+	clusterDump, err := getClusterConfigDump(configDump)
+	if err != nil {
+		return err
+	}
+	if len(clusterDump.GetDynamicActiveClusters()) == 0 {
+		return errors.New("missing dynamic active clusters")
+	}
+	return nil
+}
+
+func checkHasDynamicActiveListeners(configDump *adminapi.ConfigDump) error {
+	listenerDump, err := getListenerConfigDump(configDump)
+	if err != nil {
+		return err
+	}
+	if len(listenerDump.GetDynamicActiveListeners()) == 0 {
+		return errors.New("missing dynamic active listeners")
+	}
+	return nil
+}
+
+func getClusterConfigDump(configDump *adminapi.ConfigDump) (*adminapi.ClustersConfigDump, error) {
+	clusterDumpAny, ok := configDump.Configs[clustersMapKey]
+	if !ok {
+		return nil, fmt.Errorf("config dump has no cluster dump")
+	}
+	clusterDump := &adminapi.ClustersConfigDump{}
+	err := proto.UnmarshalAny(&clusterDumpAny, clusterDump)
+	if err != nil {
+		return nil, err
+	}
+	return clusterDump, nil
+}
+
+func getListenerConfigDump(configDump *adminapi.ConfigDump) (*adminapi.ListenersConfigDump, error) {
+	listenerDumpAny, ok := configDump.Configs[listenersMapKey]
+	if !ok {
+		return nil, fmt.Errorf("config dump has no listener dump")
+	}
+	listenerDump := &adminapi.ListenersConfigDump{}
+	err := proto.UnmarshalAny(&listenerDumpAny, listenerDump)
+	if err != nil {
+		return nil, err
+	}
+	return listenerDump, nil
+}
+
+func getConfigDump(adminPort int) (*adminapi.ConfigDump, error) {
+	requestURL := fmt.Sprintf("http://127.0.0.1:%d/config_dump", adminPort)
+	buffer, err := doHTTPGet(requestURL)
+	if err != nil {
+		return nil, err
+	}
+
+	msg := &adminapi.ConfigDump{}
+	if err := jsonpb.Unmarshal(buffer, msg); err != nil {
+		return nil, err
+	}
+	return msg, nil
+}
+
+func doHTTPGet(requestURL string) (*bytes.Buffer, error) {
+	httpClient := &http.Client{
+		Timeout: readinessTimeout,
+	}
+
+	response, err := httpClient.Get(requestURL)
+	if err != nil {
+		return nil, err
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("unexpected status %d", response.StatusCode)
+	}
+
+	var b bytes.Buffer
+	if _, err := io.Copy(&b, response.Body); err != nil {
+		return nil, err
+	}
+	return &b, nil
+}
```

Comment:
You could reuse the logic from istioctl proxy-status .. and check for sync status of all rds, cds, lds, thereby eliminating similar logic to check readiness.. @liamawhite put in a bunch of work

Reply:
ignore. see comment above

Reply:
@rshriram to be clear, you're asking to ignore your comment because istioctl proxy-status uses config_dump, correct? (the comments above are suggesting to use Envoy stats)

Reply:
Well - it may still be good to keep this into a separate package, maybe in top level /pkg - it's a general function to fetch metrics ( I think it works for prometheus metrics as well ), it could be reused at least in tests.

Reply:
@costinm you're the one who told me to move this code here in a previous review :).  It's staying.